### DISCUSSION TITLE
`remix-serve` isn't load balancer friendly

### DISCUSSION THREAD

**Discussion by joshocalico** [remix-run/remix#5196] [Proposals]
### What version of Remix are you using?

1.7.1

### Steps to Reproduce

Configure Remix to run inside an ECS container.
Setup an Application Load Balancer instance and point this at Remix.
Thrash the server, eventually (not very long after) you'll start seeing 502 errors!
This causes a failed health check from ALB and the container dies, or in this case is _savagely murdered_ by the load balancer.

### Expected Behavior

These 502s shouldn't occur. 

Remix should have an option to configure the default keepAliveTimeout.

### Actual Behavior

The socket gets closed by Remix, which then causes the load balancer to panic!

Silly brittle things they are.

**Comment by joshocalico** [comment]
For more info and a proposed solution:
- This is a known issue in NextJS too. 
https://nextjs.org/docs/api-reference/cli#keep-alive-timeout
- While I recognise this framework tends towards functions, it does support fixed compute too of course. Some non-AWS services don't have configurable upstream timeouts so it should be able to be implemented without having to use the ejector seat. i.e. a custom server.

I propose a config key and a CLI flag. 

I'll try to implement this into `remix-serve` and add a caveat to the gotchas page.

**Comment by joshocalico** [comment]
Might take me a few days as I'm pretty busy, but I'll get to it if I can. 
Feel free to swoop in if you have more time 🙂

**Comment by machour** [comment]
As if this a feature request, I'm going to convert it to a Proposal

**Reply by lucioladen** [reply]
considering that any request taking more than 5 seconds breaks connection with any LB, it would be a VERY welcomed feature. 
I know there are options to handle long-running process by forking, but even a simple external API call that took longer than usually does, could close the connection with the balancer..

**Comment by lucioladen** [comment]
I'm having the same issue. so... now I see it's not possible. 
what about using an express adapter and setting `server.keepAliveTimeout`...should do the trick right?

**Comment by chrisdostert** [comment]
seems like https://github.com/remix-run/remix/discussions/5753 is potentially a dupe or at least highly related.

**Reply by chrisdostert** [reply]
FWIW what @lucioladen recommended does work as a workaround but it would be nice to be able to use remix's **native** app server

**Reply by joshocalico** [reply]
Using the express server you do lose a few nice things remix server does for you so it's preferable as a CLI or config option

**Comment by tkhoa2711** [comment]
Is this still on the plan to be added to `remix-serve`? We're having this exact issue with where we start getting 502 errors when the traffic barely reaches 7 requests per second for our AWS load balancers (and we have 4 instances of Remix servers running!). Would be much easier to add a one-line config than having to change/adapt our code to Express.

**Comment by StevenBendell** [comment]
Is there any update on a solution here? Is the current work around of using express the only option?

**Reply by kiliman** [reply]
Remix App Server _is_ an Express server. It's just a no-config option for basic use cases.

If you need more advanced features, just use the Express template.

I have a tool that lets you easily switch from RAS to Express.

```bash
npx rmx-cli eject-ras
```

**Reply by lucioladen** [reply]
Still, being able to override configuration (talking about keepAliveTimeout along with others) through Remix config could be a generic / "one size fits all" problem solver in cases when you don't need to adapt for a different behavior but just different configuration.

**Comment by natac13** [comment]
I believe Im having a similar issue. I have a remix app dockerized and deployed to ECS with the default remix server. I put a load balancer and cloudfront in front ECS but when I go to the CDN url I first see the landing page. Then go to login and when I try to login I get redirect to the next page with only the loader json date returned and not the route component.

If changing to express fixing this then Im game but I just wanted to see if from my description I may have a different error and changing to express is not going to fix all my problems.

**Reply by natac13** [reply]
My issue was with the Distribution's caching policy

**Comment by JoaccoG** [comment]
I'm having the same issue, are there any updates? the express adapter is the only option?

**Reply by sergiodxa** [reply]
`remix-serve` is a Express server bundled for convenience, if you need something not supported there you should switch to a custom Express server, or another HTTP server.
